src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1673,6 +1673,8 @@ AssignTasksToConnections(DistributedExecution *execution)
 				WorkerSession *session =
 					FindOrCreateWorkerSession(workerPool, connection);
 
+				ConnectionStateMachine(session);
```

Comment:
I considered adding this before, but was a little hesitant since it immediately starts executing tasks and I did not know whether that could have any side effects. 

Mostly just pointing it out, might still be the right thing to do. Alternatively, maybe we should just run the connection state machine part.

Reply:
Does this have anything common with https://github.com/citusdata/citus/issues/3354 ?

Reply:
The linked issue seems to be a memory corruption issue, which this is not

Updated to only call ConnectionStateMachine when connection is CONNECTION_BAD, avoiding marco's aversion